tests/test_qnode_new.py

```diff
@@ -0,0 +1,599 @@
+    def test_calling_no_kwargs(self, operable_mock_device_2_wires):
+        """Various quantum func calling syntax errors."""
+
+        def circuit(x, y=0.2, *args, m=0.3, n):
+            circuit.in_args = (x, y, m, n)
+            return qml.expval(qml.PauliZ(0))
+
+        node = QNode(circuit, operable_mock_device_2_wires, mutable=True)
+
+        with pytest.raises(QuantumFunctionError, match="parameter 'x' given twice"):
+            node(0.1, x=1.1)
+        with pytest.raises(QuantumFunctionError, match="Unknown quantum function parameter 'foo'"):
+            node(foo=1)
+        with pytest.raises(QuantumFunctionError, match="'args' cannot be given using the keyword syntax"):
+            node(args=1)
+        with pytest.raises(QuantumFunctionError, match="positional parameter 'x' missing"):
+            node(n=0.4)
+        with pytest.raises(QuantumFunctionError, match="keyword-only parameter 'n' missing"):
+            node(0.1)
+
+        # valid calls
+        node(x=0.1, n=0.4)
+        assert circuit.in_args[2:] == (0.3, 0.4)  # first two are Variables
+        node(0.1, n=0.4)
+        assert circuit.in_args[2:] == (0.3, 0.4)
+
+    @pytest.mark.xfail(reason="Tests the keyword-only-marks-nondifferentiable syntax", raises=AssertionError, strict=True)
+    def test_calling_with_kwargs(self, operable_mock_device_2_wires):
+        """Various quantum func calling syntax errors."""
+
+        def circuit(x, y=0.2, *, m=0.3, n, **kwargs):
+            circuit.in_args = (x, y, m, n)
+            return qml.expval(qml.PauliZ(0))
+
+        node = QNode(circuit, operable_mock_device_2_wires, mutable=True)
+
+        with pytest.raises(QuantumFunctionError, match="parameter 'x' given twice"):
+            node(0.1, x=1.1)
+        with pytest.raises(QuantumFunctionError, match="'kwargs' cannot be given using the keyword syntax"):
+            node(kwargs=1)
+        with pytest.raises(QuantumFunctionError, match="takes 2 positional parameters, 3 given"):
+            node(0.1, 0.2, 100, n=0.4)
+        with pytest.raises(QuantumFunctionError, match="positional parameter 'x' missing"):
+            node(n=0.4)
+        with pytest.raises(QuantumFunctionError, match="keyword-only parameter 'n' missing"):
+            node(0.1)
+
+        # valid calls
+        node(x=0.1, n=0.4)
+        assert circuit.in_args[2:] == (0.3, 0.4)  # first two are Variables
+        node(0.1, n=0.4)
+        assert circuit.in_args[2:] == (0.3, 0.4)
+
+
+    def test_calling_bad_errors(self, operable_mock_device_2_wires):
+        """Confusing quantum func calling errors and bugs (default-value-marks-nondifferentiables syntax)."""
+
+        def circuit(x=0.1):
+            return qml.expval(qml.PauliZ(0))
+        node = QNode(circuit, operable_mock_device_2_wires)
+
+        with pytest.raises(TypeError, match="got multiple values for argument 'x'"):
+            node(0.3)  # default arg given positionally, wrong error message
```

Comment:
Same question here

Reply:
This tests the old qfunc syntax (which we still use),  a case where the old syntax leads to the wrong error message. Maybe that could be fixed though.